File: #1335.minimum-difficulty-of-a-job-schedule.cpp
class Solution {
public:
    int minDifficulty(vector<int>& jobDifficulty, int d) {

        int size = jobDifficulty.size();
        if (size < d) {
            return -1;
        }

        int dp[d][size];
        dp[0][0] = jobDifficulty[0];
        for (int i = 1; i < size; i++) {
            dp[0][i] = std::max(dp[0][i - 1],jobDifficulty[i]);
        }
        
        for (int i = 1; i < d;i++) {
            for (int j = i; j < size; j++) {
                int max = jobDifficulty[j];
                int min = INT_MAX;
                for (int k = j; k >= i; k--) {
                    max = std::max(max, jobDifficulty[k]);
                    min = std::min(min, dp[i - 1][k - 1] + max);
                }
                dp[i][j] = min;
            }
        }
        return dp[d - 1][size - 1];
    }
};

File: #1335.minimum-difficulty-of-a-job-schedule.js
/**
 * @param {number[]} jobDifficulty
 * @param {number} d
 * @return {number}
 */
var minDifficulty = function(jobDifficulty, d) {
    let n = jobDifficulty.length;
    let dp = [];
    for (let i = 0; i < n; i++) {
        dp[i] = [];
        for (let j = 1; j <= d; j++) {
            dp[i][j] = [];
        }
    }

    let diff = recur();
    if (diff === Infinity) {
        return - 1;
    }
    return diff;

    function recur(i = 0, day = 1, max = -Infinity) {
        if (i >= n) {
            return Infinity;
        }

        if (i === n - 1 && day === d) {
            return Math.max(max, jobDifficulty[i]);
        }

        if (day > d) {
            return Infinity;
        }

        if (dp[i][day][max] !== undefined) {
            return dp[i][day][max];
        }

        let currentMax = Math.max(max, jobDifficulty[i]);
        let res1 = recur(i + 1, day, currentMax);
        let res2 = currentMax + recur(i + 1, day + 1, -Infinity);
        let res = Math.min(res1, res2);
        return dp[i][day][max] = res;
    }
};
